data_steward/cdr_cleaner/cleaning_rules/null_invalid_foreign_keys.py

```diff
@@ -14,43 +14,48 @@
 
 # Python imports
 import logging
+from typing import Dict
 
 # Project Imports
 import resources
-from common import AOU_DEATH, JINJA_ENV, MAPPING_PREFIX
-from constants import bq_utils as bq_consts
-from constants.cdr_cleaner import clean_cdr as cdr_consts
+from common import AOU_DEATH, JINJA_ENV
+from constants.cdr_cleaner.clean_cdr import COMBINED, QUERY
 from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
 
 LOGGER = logging.getLogger(__name__)
 
-JIRA_ISSUE_NUMBERS = ['DC388', 'DC807']
+JIRA_ISSUE_NUMBERS = ['DC388', 'DC807', 'DC3230']
 
 FOREIGN_KEYS_FIELDS = [
     'person_id', 'visit_occurrence_id', 'location_id', 'care_site_id',
     'provider_id'
 ]
 
-INVALID_FOREIGN_KEY_QUERY = JINJA_ENV.from_string("""
-    SELECT {{cols}}
-    FROM `{{project_id}}.{{dataset_id}}.{{table_name}}` t
-    {{join_expr}}""")
-
-LEFT_JOIN = JINJA_ENV.from_string("""
-    LEFT JOIN `{{dataset_id}}.{{table}}` {{prefix}} 
-    ON t.{{field}} = {{prefix}}.{{field}}""")
-
 SANDBOX_QUERY = JINJA_ENV.from_string("""
-    CREATE OR REPLACE TABLE 
-    `{{project_id}}.{{sandbox_dataset_id}}.{{intermediary_table}}` AS (
-        SELECT t.* FROM `{{project_id}}.{{dataset_id}}.{{table_name}}` AS t 
-        WHERE {{sandbox_expr}})""")
-
-SANDBOX_EXPRESSION = JINJA_ENV.from_string("""
-     ({{field}} NOT IN (
-        SELECT {{field}} 
-        FROM `{{dataset_id}}.{{table}}` AS {{prefix}})
-        AND {{field}} IS NOT NULL)
+CREATE OR REPLACE TABLE `{{project_id}}.{{sandbox_dataset_id}}.{{sandbox_table}}` AS (
+    SELECT * FROM `{{project_id}}.{{dataset_id}}.{{table_name}}`
+    WHERE {% for key in foreign_keys %}
+        (
+            {{key}} NOT IN (
+                SELECT {{key}} FROM `{{project_id}}.{{dataset_id}}.{{key[:-3]}}`
+            )
+            AND {{key}} IS NOT NULL
+        ){% if not loop.last -%} OR {% endif %}
+    {% endfor %}
+)""")
+
+DELETE_QUERY = JINJA_ENV.from_string("""
+DELETE FROM `{{project_id}}.{{dataset_id}}.{{table_name}}`
+WHERE {{key}} NOT IN (
+    SELECT {{key}} FROM `{{project_id}}.{{dataset_id}}.{{key[:-3]}}`
```

Comment:
I see we are only sandboxing where the key is not null. This would delete NULL rows as well. Is this expected?

Reply:
Thanks for reviewing it!

This `DELETE` statement runs against `required` foreign key columns (=`person_id` for all and `visit_occurrence_id` for `visit_detail`). Since they are not nullable, we cannot update the invalid values to `NULL`, so we delete the whole row here. For `nullable` foreign key columns, we update the invalid values to `NULL`.

I added additional comments in the code. Let me know if it makes sense!

Reply:
Yes, I see what we are doing now, thanks for pointing that out and adding the comments! 

Very clear explanation as well. Thank you